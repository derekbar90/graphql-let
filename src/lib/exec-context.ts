import { ConfigTypes } from './config';
import { getCacheFullDir } from './paths';

export type ExecContext = {
  cwd: string;
  config: ConfigTypes;
  configHash: string;
  cacheFullDir: string;
};

export default function createExecContext(
  cwd: string,
  config: ConfigTypes,
  configHash: string,
) {
  const cacheFullDir = getCacheFullDir(cwd, config.cacheDir);
  return {
    cwd,
    config,
    configHash,
    cacheFullDir,
  };
}
