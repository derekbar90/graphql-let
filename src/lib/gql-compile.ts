import { generate } from '@graphql-codegen/cli';
// import { Types } from "@graphql-codegen/plugin-helpers";
import traverse from '@babel/traverse';
import makeDir from 'make-dir';
import { join as pathJoin, extname, basename, dirname } from 'path';
import { existsSync } from 'fs';
import { genDts } from './dts';
import { rimraf } from './file';
import mkdirp from 'mkdirp';
// TODO
import { createWriteStream } from 'fs';
import { stripIgnoredCharacters } from 'graphql';
import { parse, ParserOptions } from '@babel/parser';
import { readFile } from './file';
import { join } from 'path';
// import createCodegenOpts from "../../src/lib/create-codegen-opts";
import { writeFile } from './file';
import { createHash } from './hash';
// import loadConfig from "../../src/lib/load-config";
import { ConfigTypes } from './types';

const packageJsonContent = JSON.stringify({ types: 'index' }, null, 2);

export type CodegenContext = {
  gqlContent: string;
  strippedGqlContent: string;
  gqlContentHash: string;
  sourceRelPath: string;
  sourceFullPath: string;
  tsxRelPath: string;
  tsxFullPath: string;
  dtsRelPath: string;
  dtsFullPath: string;
}[];

type CacheState = {
  [hash: string]: string;
};
type CacheStateStore = {
  [tsxRelPath: string]: CacheState;
};

// const [config, configHash] = await loadConfig(cwd, configFilePath);
const config: ConfigTypes = {
  schema: 'schema/type-defs.graphqls',
  plugins: ['typescript', 'typescript-operations', 'typescript-react-apollo'],
  documents: [],
  respectGitIgnore: true,
  config: {
    reactApolloVersion: '3',
    withHOC: false,
    withHooks: true,
  },
};

const getPaths = (
  sourceRelPath: string,
  hash: string,
  dtsRelDir: string,
  libRelDir: string,
  cwd: string,
) => {
  const abs = (relPath: string) => pathJoin(cwd, relPath);

  const tsxGenFullDir = pathJoin(abs(libRelDir), '__generated__');
  const dtsGenFullDir = abs(dtsRelDir);
  // sourceRelPath: "pages/index.tsx"
  // "pages"
  const relDir = dirname(sourceRelPath);
  // ".tsx"
  const ext = extname(sourceRelPath);
  // "${cwd}/pages/index.tsx"
  const sourceFullPath = abs(sourceRelPath);
  // "index"
  const base = basename(sourceRelPath, ext);

  // "index-2345.tsx"
  const tsxBasename = `${base}-${hash}${ext}`;
  // "pages/index-2345.tsx"
  const tsxRelPath = pathJoin(relDir, tsxBasename);
  // "/Users/.../node_modules/graphql-let/__generated__/pages/index-2345.d.ts"
  const tsxFullPath = pathJoin(tsxGenFullDir, tsxRelPath);

  // "index-2345.d.ts"
  const dtsBasename = `${base}-${hash}.d.ts`;
  // "pages/index-2345.d.ts"
  const dtsRelPath = pathJoin(relDir, dtsBasename);
  // "/Users/.../node_modules/@types/graphql-let/pages/index-2345.d.ts"
  const dtsFullPath = pathJoin(dtsGenFullDir, dtsRelPath);
  // TODO
  return {
    sourceRelPath,
    sourceFullPath,
    tsxRelPath,
    tsxFullPath,
    dtsRelPath,
    dtsFullPath,
  };
};
const parserOption: ParserOptions = {
  sourceType: 'module',
  plugins: ['typescript', 'jsx'],
};

import generator from '@babel/generator';

function appendExportAsObject(dtsContent: string) {
  const exportNames: string[] = [];

  function pushExportNames({ node }: any) {
    exportNames.push(node.id.name);
  }

  const dtsAST = parse(dtsContent, parserOption);
  traverse(dtsAST, {
    ExportNamedDeclaration(path: any) {
      path.traverse({
        VariableDeclarator: pushExportNames,
        TSTypeAliasDeclaration: pushExportNames,
        FunctionDeclaration: pushExportNames,
      });
    },
    Program: {
      exit(path) {
        const pairs = exportNames.map((e) => `${e}:${e}`).join(',');
        traverse(
          parse(
            `export declare type __AllExports = { ${pairs} };`,
            parserOption,
          ),
          {
            ExportNamedDeclaration({ node }) {
              const body = path.get('body');
              body[body.length - 1].insertAfter(node);
            },
          },
        );
      },
    },
  });

  const { code } = generator(dtsAST);
  return code;
}

type GqlCompileArg = {
  cwd: string;
  dtsRelDir: string;
  libRelDir: string;
  sourceRelPath: string;
};

export async function processGqlCompile(
  cwd: string,
  dtsRelDir: string,
  libRelDir: string,
  sourceRelPath: string,
  schemaHash: string,
  gqlContents: string[],
  targetStore: CacheState,
  codegenContext: CodegenContext,
  skippedContext: CodegenContext,
) {
  /**
   * 0. Shape of storage
   * {
   *   "userRelPath.tsx": {
   *     "gqlContentHash1": "query{\n}",
   *     "gqlContentHash2": "query{\n}",
   *   }
   * }
   * 1. take care all multiple gql() in file. Check gqlContentHash and generate if not exists.
   * 2. All done. remove all old dts.
   * 3. Store the latest dts paths.
   * 4. Print index.d.ts from the entire storage
   *
   * 5. Write .tsx to cacheDir
   * 6. Import it from babel target by inserting a "import" line
   *
   * 7. Done.
   */

  for (const gqlContent of gqlContents) {
    const strippedGqlContent = stripIgnoredCharacters(gqlContent);
    const gqlContentHash = createHash(schemaHash + strippedGqlContent);
    const context = {
      gqlContent,
      strippedGqlContent,
      gqlContentHash,
      ...getPaths(sourceRelPath, gqlContentHash, dtsRelDir, libRelDir, cwd),
    };
    if (targetStore[gqlContentHash]) {
      skippedContext.push(context);
    } else {
      codegenContext.push(context);
    }
  }

  // Codegen
  if (codegenContext.length) {
    for (const { strippedGqlContent, tsxFullPath } of codegenContext) {
      const [{ content }] = await generate(
        {
          cwd,
          schema: config.schema,
          documents: [strippedGqlContent],
          generates: {
            [tsxFullPath]: {
              plugins: config.plugins,
              config: config.config,
            },
          },
        },
        false,
      );
      await mkdirp(dirname(tsxFullPath));
      await writeFile(tsxFullPath, content);
    }

    // Dts
    const dtsContents = genDts(
      codegenContext.map(({ tsxFullPath }) => tsxFullPath),
      config,
    );
    await makeDir(dirname(codegenContext[0].dtsFullPath));
    for (const [i, dtsContent] of dtsContents.entries()) {
      const {
        dtsFullPath,
        strippedGqlContent,
        gqlContentHash,
      } = codegenContext[i]!;
      targetStore[gqlContentHash] = strippedGqlContent;
      const content = appendExportAsObject(dtsContent);
      await writeFile(dtsFullPath, content);
    }
  }
}

export default async function gqlCompile(
  cwd: string,
  dtsRelDir: string,
  libRelDir: string,
  sourceRelPath: string,
  schemaHash: string,
  gqlContents: string[],
) {
  const codegenContext: CodegenContext = [];
  const skippedContext: CodegenContext = [];

  // Processes inside a sub-process of babel-plugin
  const storeFullPath = pathJoin(cwd, dtsRelDir, 'store.json');
  const store = existsSync(storeFullPath)
    ? JSON.parse(await readFile(storeFullPath, 'utf-8'))
    : {};
  const targetStore = store[sourceRelPath] || (store[sourceRelPath] = {});

  // Prepare
  await mkdirp(join(cwd, dtsRelDir));
  await writeFile(join(cwd, dtsRelDir, 'package.json'), packageJsonContent);

  await processGqlCompile(
    cwd,
    dtsRelDir,
    libRelDir,
    sourceRelPath,
    schemaHash,
    gqlContents,
    targetStore,
    codegenContext,
    skippedContext,
  );

  // Remove old caches
  for (const { gqlContentHash } of skippedContext) {
    delete targetStore[gqlContentHash];
    const { dtsFullPath } = getPaths(
      sourceRelPath,
      gqlContentHash,
      dtsRelDir,
      libRelDir,
      cwd,
    );
    if (existsSync(dtsFullPath)) {
      await rimraf(dtsFullPath);
    }
  }

  // Update index.d.ts
  const dtsEntryFullPath = pathJoin(cwd, dtsRelDir, 'index.d.ts');
  const writeStream = createWriteStream(dtsEntryFullPath);
  for (const { gqlContent, gqlContentHash, dtsRelPath } of codegenContext) {
    const chunk = `import T${gqlContentHash} from './${dtsRelPath}';
export declare function gql(gql: \`${gqlContent}\`): T${gqlContentHash}.__AllExports;
`;
    await new Promise((resolve) => writeStream.write(chunk, resolve));
  }

  // Update storeJson
  await writeFile(storeFullPath, JSON.stringify(store, null, 2));
}
