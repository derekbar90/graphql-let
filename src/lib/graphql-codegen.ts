import { Types } from '@graphql-codegen/plugin-helpers';
import { generate } from '@graphql-codegen/cli';
import makeDir from 'make-dir';
import path from 'path';
import { ExecContext } from './exec-context';
import { withHash, writeFile } from './file';
import { printError } from './print';
import {
  CodegenContext,
  FileCodegenContext,
  isLiteralContext,
  LiteralCodegenContext,
} from './types';
import { CodegenContext as CodegenConfig } from '@graphql-codegen/cli';
import ConfiguredOutput = Types.ConfiguredOutput;

export function buildCodegenConfig(
  { cwd, config }: ExecContext,
  codegenContext: CodegenContext[],
) {
  const generates: {
    [outputPath: string]: ConfiguredOutput;
  } = Object.create(null);

  for (const context of codegenContext) {
    const { tsxFullPath } = context;
    const documents = isLiteralContext(context)
      ? // XXX: We want to pass shorter `strippedGqlContent`,
        // but `# import` also disappears!
        (context as LiteralCodegenContext).gqlContent
      : (context as FileCodegenContext).gqlRelPath;
    generates[tsxFullPath] = {
      ...config.generateOptions,
      // graphql-let -controlled fields:
      documents,
      plugins: config.plugins,
    };
  }

  return {
    ...config,
    // @ts-ignore
    cwd,

    // @ts-ignore This allows recognizing "#import" in GraphQL documents
    skipGraphQLImport: false,

    // In our config, "documents" should always be empty
    // since "generates" should take care of them.
    documents: undefined,
    generates,
  };
}

export async function processGraphQLCodegen(
  execContext: ExecContext,
  codegenContext: CodegenContext[],
  generateArg: CodegenConfig | (Types.Config & { cwd?: string }),
): Promise<Types.FileOutput[]> {
  let results: Types.FileOutput[];
  try {
    results = await generate(generateArg, false);
  } catch (error) {
    if (error.name === 'ListrError' && error.errors != null) {
      for (const e of error.errors) printError(e);
      throw error.errors[0];
    } else {
      console.log('Error:', error);
    }
    throw error;
  }
  if (codegenContext.length !== results.length) throw new Error('never');
  // Object option "generates" in codegen obviously doesn't guarantee result's order.
  const tsxPathTable = new Map<string, CodegenContext>(
    codegenContext.map((c) => [c.tsxFullPath, c]),
  );
  for (const result of results) {
    const { filename, content } = result;
    const context = tsxPathTable.get(filename);
    if (!context) throw new Error('never');
    await makeDir(path.dirname(filename));
    await writeFile(filename, withHash(context.gqlHash, content));
  }
  return results;
}
