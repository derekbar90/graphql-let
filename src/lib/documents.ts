import { CodegenContext as CodegenConfig } from '@graphql-codegen/cli';
import { Types } from '@graphql-codegen/plugin-helpers';
import { GraphQLFileLoader } from '@graphql-tools/graphql-file-loader';
import glob from 'globby';
import { join as pathJoin } from 'path';
import { ConfigTypes } from './config';
import { ExecContext } from './exec-context';
import { readFile, readHash } from './file';
import { processGraphQLCodegen } from './graphql-codegen';
import { createHash } from './hash';
import { createPaths, isTypeScriptPath } from './paths';
import {
  CodegenContext,
  FileCodegenContext,
  isLiteralContext,
  LiteralCodegenContext,
} from './types';
import ConfiguredOutput = Types.ConfiguredOutput;

export async function findTargetDocuments({
  cwd,
  config,
}: ExecContext): Promise<{
  graphqlRelPaths: string[];
  tsSourceRelPaths: string[];
}> {
  const documentPaths = await glob(config.documents, {
    cwd,
    gitignore: config.respectGitIgnore,
  });
  if (documentPaths.length === 0) {
    throw new Error(
      `No GraphQL documents are found from the path ${JSON.stringify(
        config.documents,
      )}. Check "documents" in .graphql-let.yml.`,
    );
  }
  const graphqlRelPaths: string[] = [];
  const tsSourceRelPaths: string[] = [];
  for (const p of documentPaths) {
    isTypeScriptPath(p) ? tsSourceRelPaths.push(p) : graphqlRelPaths.push(p);
  }
  return { graphqlRelPaths, tsSourceRelPaths };
}

type ProcessDocumentsForContextReturnType = Record<
  /*gqlRelPath*/ string,
  /*tsxContent*/ string
>;

class GraphQLLetConfig extends CodegenConfig {
  constructor(execContext: ExecContext, codegenContext: CodegenContext[]) {
    const { cwd, config, codegenOpts } = execContext;

    // In our config, "documents" should always be empty
    // since "generates" should take care of them.
    const generates = GraphQLLetConfig.buildGenerates(
      execContext,
      codegenContext,
      config,
    );
    super({
      config: {
        ...config,
        // @ts-ignore
        cwd,
        config: {
          // TODO: Quit using codegenOpts
          ...codegenOpts.config,
          ...config.config,
        },
        // schema: path.join(cwd, config.schema as any),
        documents: undefined,
        generates,
      },
    });
    this.cwd = cwd;
  }

  static buildGenerates(
    execContext: ExecContext,
    codegenContext: CodegenContext[],
    config: Omit<Types.Config, 'generates'> & ConfigTypes,
  ) {
    const generates: {
      [outputPath: string]: ConfiguredOutput;
    } = Object.create(null);
    for (const context of codegenContext) {
      const { tsxFullPath } = context;
      const documents = isLiteralContext(context)
        ? (context as LiteralCodegenContext).strippedGqlContent
        : (context as FileCodegenContext).gqlRelPath;
      generates[tsxFullPath] = {
        ...config.generateOptions,
        // graphql-let -controlled fields:
        documents,
        plugins: config.plugins,
      };
    }
    return generates;
  }

  // from graphql-file-loader
  static isGraphQLImportFile(rawSDL: string) {
    const trimmedRawSDL = rawSDL.trim();
    return (
      trimmedRawSDL.startsWith('# import') ||
      trimmedRawSDL.startsWith('#import')
    );
  }

  async loadDocuments(pointers: any) {
    const [pointer] = pointers;
    if (GraphQLLetConfig.isGraphQLImportFile(pointer)) {
      // GraphQLFileLoader only allows "# import" when passing file paths.
      // But we want it even in gql(`query {}`), don't we?
      const resolved = GraphQLFileLoader.prototype.handleFileContent(
        pointer,
        './a.graphql',
        { cwd: this.cwd },
      );
      return [resolved];
    }
    return super.loadDocuments(pointers); // , this.getConfig());
  }
}

export function processGraphQLCodegenForDocuments(
  execContext: ExecContext,
  codegenContext: CodegenContext[],
) {
  return processGraphQLCodegen(
    execContext,
    codegenContext,
    new GraphQLLetConfig(execContext, codegenContext),
  );
}

export async function processDocumentsForContext(
  execContext: ExecContext,
  schemaHash: string,
  codegenContext: CodegenContext[],
  gqlRelPaths: string[],
  gqlContents?: string[],
) {
  const tsxContents: ProcessDocumentsForContextReturnType = Object.create(null);
  if (!gqlRelPaths.length) return tsxContents;

  const documentCodegenContext: CodegenContext[] = [];

  const { cwd } = execContext;
  for (const [i, gqlRelPath] of gqlRelPaths.entries()) {
    // Loader passes gqlContent directly
    const gqlContent = gqlContents
      ? gqlContents[i]
      : await readFile(pathJoin(cwd, gqlRelPath), 'utf-8');
    if (!gqlContent) throw new Error('never');

    const createdPaths = createPaths(execContext, gqlRelPath);
    const { tsxFullPath, dtsFullPath } = createdPaths;

    // Here I add "schemaHash" as a hash seed. Types of GraphQL documents
    // basically depends on schema, which change should effect to document results.
    const gqlHash = createHash(schemaHash + gqlContent);

    const shouldUpdate =
      gqlHash !== (await readHash(tsxFullPath)) ||
      gqlHash !== (await readHash(dtsFullPath));

    const context: FileCodegenContext = {
      ...createdPaths,
      gqlHash,
      dtsContentDecorator: (s) => s,
      skip: !shouldUpdate,
    };
    codegenContext.push(context);
    documentCodegenContext.push(context);
  }
  const results = await processGraphQLCodegenForDocuments(
    execContext,
    documentCodegenContext,
  );
  for (const [i, { content }] of results.entries()) {
    const { gqlRelPath } = codegenContext[i]! as FileCodegenContext;
    tsxContents[gqlRelPath] = content;
  }
  return tsxContents;
}
