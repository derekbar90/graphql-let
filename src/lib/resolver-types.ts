import { extname, join as pathJoin } from 'path';
import { ConfigTypes } from './config';
import { ExecContext } from './exec-context';
import { readFile, readHash } from './file';
import { createHash } from './hash';
import { createPaths } from './paths';
import { PRINT_PREFIX, updateLog } from './print';
import { processGraphQLCodegen } from './graphql-codegen';
import { CodegenContext, FileCodegenContext } from './types';

// Currently glob for schema is not allowed.
function isLocalFilePathWithExtension(s: any): boolean {
  if (typeof s !== 'string') return false;
  if (extname(s).length) return true;
  return false;
}

export function shouldGenResolverTypes(config: ConfigTypes): boolean {
  try {
    require('@graphql-codegen/typescript');
    require('@graphql-codegen/typescript-resolvers');

    if (isLocalFilePathWithExtension(config.schema)) return true;
    console.info(
      PRINT_PREFIX +
        'Failed to generate Resolver Types. You have to specify at least one schema (glob) path WITH an extension, such as "**/*.graphqls"',
    );
    return false;
  } catch (e) {
    // Just skip.
    return false;
  }
}

export async function prepareGenResolverTypes(execContext: ExecContext) {
  const { cwd, config, configHash } = execContext;
  const fileSchema = config.schema as string;
  const schemaFullPath = pathJoin(cwd, fileSchema);
  const content = await readFile(schemaFullPath);
  const schemaHash = createHash(configHash + content);
  return { schemaFullPath, schemaHash };
}

export async function processResolverTypesIfNeeded(
  execContext: ExecContext,
  codegenContext: CodegenContext[],
) {
  const { cwd, config, configHash } = execContext;
  // To pass config change on subsequent generation,
  // configHash should be primary hash seed.
  let schemaHash = configHash;

  if (shouldGenResolverTypes(config)) {
    const fileSchema = config.schema as string;
    const schemaFullPath = pathJoin(cwd, fileSchema);
    const content = await readFile(schemaFullPath);
    schemaHash = createHash(schemaHash + content);
    const createdPaths = createPaths(execContext, fileSchema);

    const shouldUpdate =
      schemaHash !== (await readHash(createdPaths.tsxFullPath)) ||
      schemaHash !== (await readHash(createdPaths.dtsFullPath));

    const context: FileCodegenContext = {
      ...createdPaths,
      gqlHash: schemaHash,
      dtsContentDecorator: (s) => {
        return `${s}
          
// This is an extra code in addition to what graphql-codegen makes.
// Users are likely to use 'graphql-tag/loader' with 'graphql-tag/schema/loader'
// in webpack. This code enables the result to be typed.
import { DocumentNode } from 'graphql'
export default typeof DocumentNode
`;
      },
      skip: !shouldUpdate,
    };
    codegenContext.push(context);

    if (shouldUpdate) {
      // We don't delete tsxFullPath and dtsFullPath here because:
      // 1. We'll overwrite them so deleting is not necessary
      // 2. Windows throws EPERM error for the deleting and creating file process.

      updateLog(
        `Local schema files are detected. Generating resolver types...`,
      );

      await processGraphQLCodegen(execContext, [context], {
        ...config,
        cwd,
        generates: {
          [context.tsxFullPath]: {
            plugins: ['typescript', 'typescript-resolvers'],
          },
        },
      });
    }
  }
  return { schemaHash };
}
