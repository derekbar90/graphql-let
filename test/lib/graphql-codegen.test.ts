import { join as pathJoin } from 'path';
import loadConfig from '../../src/lib/config';
import { processGraphQLCodegenForFiles } from '../../src/lib/documents';
import createExecContext from '../../src/lib/exec-context';
import { createPaths } from '../../src/lib/paths';
import { FileCodegenContext } from '../../src/lib/types';
import { matchPathsAndContents } from '../__tools/match-paths-and-contents';

const cwd = pathJoin(__dirname, '../__fixtures/graphql-codegen');

describe('graphql-codegen.ts', () => {
  test('works', async () => {
    const [config, configHash] = await loadConfig(cwd);

    const execContext = createExecContext(cwd, config, configHash);
    const paths = createPaths(execContext, 'pages/viewer.graphql');
    const codegenContext: FileCodegenContext[] = [
      {
        ...paths,
        gqlHash: 'xx',
        skip: false,
        dtsContentDecorator: (_) => _,
      },
    ];
    await processGraphQLCodegenForFiles(execContext, codegenContext);
    await matchPathsAndContents(['__generated__'], cwd);
  });
});
