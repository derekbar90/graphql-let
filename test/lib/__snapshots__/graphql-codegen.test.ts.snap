// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`graphql-codegen.ts works: __generated__ 1`] = `
Array [
  "__generated__/pages/viewer.graphql.tsx",
]
`;

exports[`graphql-codegen.ts works: __generated__/pages/viewer.graphql.tsx 1`] = `
"/* xx
 * This file is automatically generated by graphql-let. */

import gql from 'graphql-tag';
import * as React from 'react';
import * as ApolloReactCommon from '@apollo/react-common';
import * as ApolloReactComponents from '@apollo/react-components';
import * as ApolloReactHooks from '@apollo/react-hooks';
export type Maybe<T> = T | null;
export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] };
export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
};

export type Query = {
  __typename?: 'Query';
  viewer?: Maybe<User>;
};

export type User = {
  __typename?: 'User';
  id: Scalars['ID'];
  name: Scalars['String'];
  status: Scalars['String'];
};

export type ViewerQueryVariables = Exact<{ [key: string]: never; }>;


export type ViewerQuery = (
  { __typename?: 'Query' }
  & { viewer?: Maybe<(
    { __typename?: 'User' }
    & Pick<User, 'id' | 'name' | 'status'>
  )> }
);


export const ViewerDocument = gql\`
    query Viewer {
  viewer {
    id
    name
    status
  }
}
    \`;
export type ViewerComponentProps = Omit<ApolloReactComponents.QueryComponentOptions<ViewerQuery, ViewerQueryVariables>, 'query'>;

    export const ViewerComponent = (props: ViewerComponentProps) => (
      <ApolloReactComponents.Query<ViewerQuery, ViewerQueryVariables> query={ViewerDocument} {...props} />
    );
    

/**
 * __useViewerQuery__
 *
 * To run a query within a React component, call \`useViewerQuery\` and pass it any options that fit your needs.
 * When your component renders, \`useViewerQuery\` returns an object from Apollo Client that contains loading, error, and data properties
 * you can use to render your UI.
 *
 * @param baseOptions options that will be passed into the query, supported options are listed on: https://www.apollographql.com/docs/react/api/react-hooks/#options;
 *
 * @example
 * const { data, loading, error } = useViewerQuery({
 *   variables: {
 *   },
 * });
 */
export function useViewerQuery(baseOptions?: ApolloReactHooks.QueryHookOptions<ViewerQuery, ViewerQueryVariables>) {
        return ApolloReactHooks.useQuery<ViewerQuery, ViewerQueryVariables>(ViewerDocument, baseOptions);
      }
export function useViewerLazyQuery(baseOptions?: ApolloReactHooks.LazyQueryHookOptions<ViewerQuery, ViewerQueryVariables>) {
          return ApolloReactHooks.useLazyQuery<ViewerQuery, ViewerQueryVariables>(ViewerDocument, baseOptions);
        }
export type ViewerQueryHookResult = ReturnType<typeof useViewerQuery>;
export type ViewerLazyQueryHookResult = ReturnType<typeof useViewerLazyQuery>;
export type ViewerQueryResult = ApolloReactCommon.QueryResult<ViewerQuery, ViewerQueryVariables>;"
`;
